xapian-core/weight/weight.cc

```diff
@@ -134,4 +144,25 @@ Weight::unserialise(const string &) const
     throw Xapian::UnimplementedError("unserialise() not supported for this Xapian::Weight subclass");
 }
 
+const Weight *
+Weight::create(const string & s, const Registry & reg)
+{
+    const char *p = s.c_str();
+    std::string scheme;
+
+    while (*p != ' ') {
+	if (*p == '\0') break;
+	scheme += *p;
+	p++;
+    }
+
+    return reg.get_weighting_scheme(scheme)->create_from_parameters(p);
```

Comment:
We should increment `p` here if it was a space rather than the end of the string.